pytorch_tabnet/abstract_model.py

```diff
@@ -320,177 +315,366 @@ def load_model(self, filepath):
+
+        return scores
+
+    def _set_network(self):
+        """Setup the network and explain matrix."""
+        self.network = tab_network.TabNet(
+            self.input_dim,
+            self.output_dim,
+            n_d=self.n_d,
+            n_a=self.n_a,
+            n_steps=self.n_steps,
+            gamma=self.gamma,
+            cat_idxs=self.cat_idxs,
+            cat_dims=self.cat_dims,
+            cat_emb_dim=self.cat_emb_dim,
+            n_independent=self.n_independent,
+            n_shared=self.n_shared,
+            epsilon=self.epsilon,
+            virtual_batch_size=self.virtual_batch_size,
+            momentum=self.momentum,
+            device_name=self.device_name,
+            mask_type=self.mask_type,
+        ).to(self.device)
+
+        self.reducing_matrix = create_explain_matrix(
+            self.network.input_dim,
+            self.network.cat_emb_dim,
+            self.network.cat_idxs,
+            self.network.post_embed_dim,
+        )
+
+    def _set_metrics(self, metrics, eval_names):
+        """Set attributes relative to the metrics.
 
         Parameters
         ----------
-            data: a :tensor: `torch.Tensor`
-                Input data
-            target: a :tensor: `torch.Tensor`
-                Target data
+        metrics : list of str
+            List of eval metric names.
+        eval_names : list of str
+            List of eval set names.
 
-        Returns
-        -------
-            predictions: np.array
-                Predictions of the regression problem or the last class
         """
-        raise NotImplementedError('users must define predict to use this base class')
+        metrics = metrics or [self._default_metric]
+
+        metrics = check_metrics(metrics)
+        # Set metric container for each sets
+        self._metric_container_dict = {
+            "train": MetricContainer(metrics, task=self._task, prefix="train_")
+        }
+        for name in eval_names:
+            self._metric_container_dict.update(
+                {name: MetricContainer(metrics, task=self._task, prefix=f"{name}_")}
+            )
+
+        self._metrics = []
+        self._metrics_names = []
+        for _, metric_container in self._metric_container_dict.items():
+            self._metrics.extend(metric_container.metrics)
+            self._metrics_names.extend(metric_container.names)
+
+        # Early stopping metric is the last eval metric
+        self.early_stopping_metric = self._metrics_names[-1]
+
+    def _set_callbacks(self, custom_callbacks):
+        """Setup the callbacks functions.
+
+        Parameters
+        ----------
+        callbacks : list of func
+            List of callback functions.
 
-    def explain(self, X):
         """
-        Return local explanation
+        # Setup default callbacks history and early stopping
+        self.history = History(self, verbose=self.verbose)
+        early_stopping = EarlyStopping(
```

Comment:
I think if no eval_set is given we should not use early stopping on the training set
```suggestion
        callbacks = [self.history]
        if self.eval_set is not None:
            early_stopping = EarlyStopping(...)
            callbacks.append(early_stopping)
```
What do you think?

Reply:
yes, done